File: tests/algorithms/x/Java/maths/series/geometric.java
public class Main {

    static boolean is_geometric_series(double[] series) {
        if (series.length == 0) {
            throw new RuntimeException(String.valueOf("Input list must be a non empty list"));
        }
        if (series.length == 1) {
            return true;
        }
        if (series[0] == 0.0) {
            return false;
        }
        double ratio = series[1] / series[0];
        int i = 0;
        while (i < series.length - 1) {
            if (series[i] == 0.0) {
                return false;
            }
            if (series[i + 1] / series[i] != ratio) {
                return false;
            }
            i = i + 1;
        }
        return true;
    }

    static double geometric_mean(double[] series) {
        if (series.length == 0) {
            throw new RuntimeException(String.valueOf("Input list must be a non empty list"));
        }
        double product = 1.0;
        int i_1 = 0;
        while (i_1 < series.length) {
            product = product * series[i_1];
            i_1 = i_1 + 1;
        }
        int n = series.length;
        return nth_root(product, n);
    }

    static double pow_float(double base, int exp) {
        double result = 1.0;
        int i_2 = 0;
        while (i_2 < exp) {
            result = result * base;
            i_2 = i_2 + 1;
        }
        return result;
    }

    static double nth_root(double value, int n) {
        if (value == 0.0) {
            return 0.0;
        }
        double low = 0.0;
        double high = value;
        if (value < 1.0) {
            high = 1.0;
        }
        double mid = (low + high) / 2.0;
        int i_3 = 0;
        while (i_3 < 40) {
            double mp = pow_float(mid, n);
            if (mp > value) {
                high = mid;
            } else {
                low = mid;
            }
            mid = (low + high) / 2.0;
            i_3 = i_3 + 1;
        }
        return mid;
    }

    static void test_geometric() {
        double[] a = ((double[])(new double[]{2.0, 4.0, 8.0}));
        if (!(Boolean)is_geometric_series(((double[])(a)))) {
            throw new RuntimeException(String.valueOf("expected geometric series"));
        }
        double[] b = ((double[])(new double[]{1.0, 2.0, 3.0}));
        if (((Boolean)(is_geometric_series(((double[])(b)))))) {
            throw new RuntimeException(String.valueOf("expected non geometric series"));
        }
    }

    static void main() {
        test_geometric();
        System.out.println(geometric_mean(((double[])(new double[]{2.0, 4.0, 8.0}))));
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            main();
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }
}

File: tests/algorithms/transpiler/FS/maths/series/geometric.fs
// Generated 2025-08-08 18:09 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _dictAdd<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) (v:'V) =
    d.[k] <- v
    d
let _dictCreate<'K,'V when 'K : equality> (pairs:('K * 'V) list) : System.Collections.Generic.IDictionary<'K,'V> =
    let d = System.Collections.Generic.Dictionary<'K, 'V>()
    for (k, v) in pairs do
        d.[k] <- v
    upcast d
let _dictGet<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) : 'V =
    match d.TryGetValue(k) with
    | true, v -> v
    | _ -> Unchecked.defaultof<'V>
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec is_geometric_series (series: float array) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable series = series
    try
        if (Seq.length (series)) = 0 then
            failwith ("Input list must be a non empty list")
        if (Seq.length (series)) = 1 then
            __ret <- true
            raise Return
        if (_idx series (0)) = 0.0 then
            __ret <- false
            raise Return
        let ratio: float = (_idx series (1)) / (_idx series (0))
        let mutable i: int = 0
        while i < ((Seq.length (series)) - 1) do
            if (_idx series (i)) = 0.0 then
                __ret <- false
                raise Return
            if ((_idx series (i + 1)) / (_idx series (i))) <> ratio then
                __ret <- false
                raise Return
            i <- i + 1
        __ret <- true
        raise Return
        __ret
    with
        | Return -> __ret
and geometric_mean (series: float array) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable series = series
    try
        if (Seq.length (series)) = 0 then
            failwith ("Input list must be a non empty list")
        let mutable product: float = 1.0
        let mutable i: int = 0
        while i < (Seq.length (series)) do
            product <- product * (_idx series (i))
            i <- i + 1
        let n: int = Seq.length (series)
        __ret <- nth_root (product) (n)
        raise Return
        __ret
    with
        | Return -> __ret
and pow_float (``base``: float) (exp: int) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable ``base`` = ``base``
    let mutable exp = exp
    try
        let mutable result: float = 1.0
        let mutable i: int = 0
        while i < exp do
            result <- result * ``base``
            i <- i + 1
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
and nth_root (value: float) (n: int) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable value = value
    let mutable n = n
    try
        if value = 0.0 then
            __ret <- 0.0
            raise Return
        let mutable low: float = 0.0
        let mutable high: float = value
        if value < 1.0 then
            high <- 1.0
        let mutable mid: float = (low + high) / 2.0
        let mutable i: int = 0
        while i < 40 do
            let mp: float = pow_float (mid) (n)
            if mp > value then
                high <- mid
            else
                low <- mid
            mid <- (low + high) / 2.0
            i <- i + 1
        __ret <- mid
        raise Return
        __ret
    with
        | Return -> __ret
and test_geometric () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let a: float array = unbox<float array> [|2.0; 4.0; 8.0|]
        if not (is_geometric_series (a)) then
            failwith ("expected geometric series")
        let b: float array = unbox<float array> [|1.0; 2.0; 3.0|]
        if is_geometric_series (b) then
            failwith ("expected non geometric series")
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        test_geometric()
        printfn "%g" (geometric_mean (unbox<float array> [|2.0; 4.0; 8.0|]))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
